コメントを付けられるようにする

takao0206 opened the pull request:
プラクティスを提出致します。ご確認よろしくお願い致します。

takao0206 committed:
日報(reports)の基本的な機能を作成。

takao0206 committed:
ユーザは自分自身が作成した日報だけを編集、削除できる機能を作成。

takao0206 committed:
日報にコメントをつけられる機能を作成。

takao0206 committed:
本にコメントをつけられる機能を作成。

takao0206 committed:
application.cssファイルに改行コードを追加。

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,59 @@
+    if @comment.save
+      redirect_to @commentable, notice: t('controllers.common.notice_create', name: Comment.model_name.human)
+    else
+      redirect_to @commentable, alert: t('controllers.common.alert_create', name: Comment.model_name.human)
```
ここで `redirect_to` すると、バリデーションエラーが発生した時にユーザーが入力していたコメントが消えてしまったり、 `@comment.errors` の内容が消えてしまい適切なトラブルシュートの内容を画面に出しにくかったり、といった問題があります。
Rails wayに従って `render` することは難しいでしょうか？

takao0206 replied:
リダイレクトすることでエラー情報が消えてしまう問題点に気付いておりませんでした。アドバイスありがとうございました。
コメントを編集するフォーム (:edit)をレンダリングするように変更したいと思います。

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,59 @@
+  end
+
+  def find_commentable
+    if params[:book_id]
```
考え方としてはこれで合っています 🙆‍♂️ 
ただ、この方法だと今後の仕様追加でコメントできるものが増えるに従って、このメソッド・クラスも複雑化していく懸念があります。

CommentControllerを継承したコントローラをコメント対象ごとに作成して、コメント対象に応じて挙動が変わる部分のみ子クラスに記載する、といったことができないか検討をお願いします 🙏

takao0206 replied:
アドバイスありがとうございました。名前空間を使ってメソッド`find_commentable`を`reports`用と`books`用に分けた子クラスを作成（CommentControllerを継承）してみたいと思います。

yuuu commented on app/helpers/comments_helper.rb:
```diff
@@ -0,0 +1,7 @@
+
+module CommentsHelper
+  def new_comment_for(commentable)
+    current_user.comments.build(commentable:)
```
この処理はHelperというよりコントローラで行うべきですね。

takao0206 replied:
アドバイス、ありがとうございました。ヘルパーではなくコントローラーで処理するように変更したいと思います。

yuuu commented on app/views/reports/show.html.erb:
```diff
@@ -0,0 +1,53 @@
+      <ul>
+        <% @report.comments.each do |comment| %>
+          <li class="comment-list">
+            <strong><%= comment.user.name.presence || comment.user.email %></strong> - <%= l comment.created_at, format: :long %>
```
`name` or `email` のどちらを表示するか決める処理ですが、これは `User` のメソッドとすることでView上の分岐を減らしたいですね。

takao0206 replied:
アドバイス、ありがとうございました。モデル`User`のメソッドにすることで、viewsはシンプルになり、メソッドの流用も容易になると思います。

takao0206 committed:
コメントを保存時にエラーが発生した場合、リダイレクトではなく、編集フォームにレンダリングするように変更。

takao0206 committed:
commentのインスタンス作成を、ヘルパーでなく、コントローラーで作成するように変更。

takao0206 committed:
名前空間を使って、`comments_controller`内のprivateメソッド`find_commentable`を`reports`用と`books`用に分けた子クラスを作成。

takao0206 committed:
データベースのレコード確認のための方法を、`any?`から`exists?`に変更。

takao0206 committed:
ユーザー名として`name`か`email`を表示する処理を、viewsからmodels（`User`のメソッド）へ移行。

takao0206 commented:
コード修正のご回答が遅くなりました。申し訳ございませんでした。
ご都合の良いときにご確認いただければ幸いです。
よろしくお願い致します。

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,51 @@
+    if @comment.save
+      redirect_to @commentable, notice: t('controllers.common.notice_create', name: Comment.model_name.human)
+    else
+      render :edit, status: :unprocessable_entity
```
ここは `create` メソッドなので `render` 先は `:edit` ではないのではないでしょうか？

takao0206 replied:
ご指摘、ありがとうございます。
booksとreportsそれぞれの`:show`にレンダリングするように修正したいと思います。
（ここにも名前空間を使ってみたいと思います）

yuuu commented on app/views/comments/_comments_section.html.erb:
```diff
@@ -0,0 +1,34 @@
+  <h2><%= t('views.common.title_index', name: Comment.model_name.human) %></h2>
+  <% if commentable.comments.any? %>
+    <ul>
+      <% commentable.comments.each do |comment| %>
```
ここは明示的に `order` をしないと順番がランダムになる可能性がありますね。

takao0206 replied:
アドバイス、ありがとうございます。
`order`を入れる方法を調査して、修正したいと思います。

yuuu commented on app/views/comments/_comments_section.html.erb:
```diff
@@ -0,0 +1,34 @@
+<div>
+  <h2><%= t('views.common.title_index', name: Comment.model_name.human) %></h2>
+  <% if commentable.comments.any? %>
```
ここは `any?` を使うよりも `exists?` を使ったほうが、SQLに `EXISTS句` を渡せるため効率が良いです。

takao0206 replied:
アドバイス、ありがとうございます。
SQLでexistsを使うと、レコードの存在有無を確認し、不要なデータの取得を行わないので、効率がいいのですね。勉強になりました。
修正したいと思います。

takao0206 committed:
コメントの一覧の表示を日付順に変更（`order(created_at: :asc)`を使用）。

takao0206 committed:
コメント作成中にエラーが出た場合、booksとreportsそれぞれの`:show`にレンダリングするように修正。

takao0206 committed:
Merge branch 'my-comments' of https://github.com/takao0206/fjord-books_app-2023 into my-comments

takao0206 commented:
コードを修正致しました。
ご都合の良いときにご確認いただければ幸いです。

yuuu commented on app/controllers/books/comments_controller.rb:
```diff
@@ -1,6 +1,20 @@
+    else
+      flash[:alert] = @comment.errors.full_messages.first
+      @book = @commentable
+      render 'books/show', status: :unprocessable_entity
```
BooksとReportsで切り替えたいのは `render` の第1引数のみのはずなので、それ以外の処理は親クラスに配置してはどうでしょうか？

takao0206 replied:
アドバイスいただき、ありがとうございます。
おっしゃるとおり、変化しているところ以外は、共通（親クラス）にしておいた方が、コードがシンプルになると思います。勉強になりました。
修正したいと思います。

takao0206 committed:
commentsのコントローラーで、`create`メソッドの共通部分を親クラスに移動。

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,65 @@
+
+  def destroy
+    @commentable = @comment.commentable
+    @comment.destroy
```
ここですが、DBに以上があるときや何らかの理由でコメントの削除に失敗したときに、きちんと例外が上がってくるように `destroy!` を使ったほうが良いですね。
日報についても同様にお願いします 🙏

takao0206 replied:
`destroy!`のご説明、ありがとうございます。
`destroy`と`destroy!`の違いについて、勉強することができました。
例外は表示できた方が良いと思いますので、コメントと日報のコントローラーについて`destroy!`に修正したいと思います。

takao0206 committed:
`reports`と`comments`のコントローラーの削除機能を`destroy`から`destroy!`に変更。

- ルーティングも修正。コメント削除時のレンダリング先を明確にする（「本」か「日報」かを正しく選ぶ）ために、ルーティングの`shallow: true`を外し、URLに`books`や`reports`が入るようにした。

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -31,8 +31,16 @@ def update
+    rescue ActiveRecord::RecordNotDestroyed => e
+      flash[:alert] = e.message
+      set_commentable
+      @comment = Comment.new
```
ここに到達する時点で `@comment` はセットされているのでこの行は不要ではないでしょうか？

takao0206 replied:
早速のコメント、ありがとうございました。

おっしゃる通り、`@comment`はすでにセットされていました。
その内容がビュー（`views/comments/_comments_section.html.erb`）に渡されて「コメント新規作成」に表示されていました。
そこで、`@comment = Comment.new`を挿入して、コメント新規作成の欄が空になるようにしていました。
`@comment = Comment.new`は必須ではありませんので、削除したいと思います。

takao0206 committed:
コメント削除時の例外処理に関するコードの修正（`@comment = Comment.new`の削除等）
